File searching bug

kbwestfall opened the pull request:
This does a few minor things:

 - Brings `develop` up to date with `release` after the 1.14.0 tag
 - Fixes a bug in the error message when the code finds more than one file with the same name (but presumably different extensions).
 - I'd like us to deprecate use of CHANGES and instead directly add to new "release" docs.  Each time we do a release, I consolidate the listed changes into a doc and add it to our "What's New" page.  I'm asking that we do that as we go now.  If your change doesn't really fit under one of the existing headings, please add a new one.  And, for each change, provide a narrative sentence or two explaining the change in a way that other developers (and eventually users) can understand.  Other suggestions welcome.

kbwestfall committed:
fix file search

kbwestfall committed:
rm glob

kbwestfall committed:
CHANGES + docs

kbwestfall committed:
missed release doc

tbowers7 commented on CHANGES.rst:
```diff
@@ -1,6 +1,13 @@
+
+1.14.0 (18 Sep 2023)
+--------------------
+
```
This is a really great idea -- and decreases the workload associated with tagging a release.

tbowers7 commented on pypeit/spectrographs/jwst_nirspec.py:
```diff
@@ -288,11 +288,9 @@ def get_rawimage(self, raw_file, det):
-        if len(fil) != 1:
-            msgs.error("Found {:d} files matching {:s}".format(len(fil)))
-
+        fil = utils.find_single_file(f'{raw_file}*')
+        if fil is None:
+            msgs.error(f'No files matching pattern: {raw_file}')
```
This is a nice replacement.

In `utils.find_single_file()`, it might be helpful to indicate to the user _which_ file is being used in the event of multiple matches (current warning just says "using the first one").

kbwestfall replied:
Good suggestion!  Done.

tbowers7 commented on doc/releases/1.14.1dev.rst:
```diff
@@ -0,0 +1,29 @@
+Version 1.14.1dev
+=================
+
+Dependency Changes
+------------------
+
+Functionality/Performance Improvements and Additions
+----------------------------------------------------
+
+Instrument-specific Updates
+---------------------------
+
+Script Changes
+--------------
+
+Datamodel Changes
+-----------------
+
+Under-the-hood Improvements
+---------------------------
+
+Bug Fixes
+---------
```
Having all of these headings for the developer to use is very helpful (rather than the semi-chronological order in CHANGES.rst).

tbowers7 reviewed (approved):
Great stuff.  One minor suggestion about adding to `utils.find_single_file()`, but approving now.

kbwestfall committed:
name file returned by find_single_file

kbwestfall committed:
minor function clean-up

kbwestfall committed:
minor function clean-up

kbwestfall committed:
list bug fix

tbowers7 commented on pypeit/utils.py:
```diff
@@ -1790,24 +1790,22 @@ def is_float(s):
+        one.
     """
-
     files = glob.glob(file_pattern)
```
Not sure if it matters, but if we sort this list, there might be a consistent pattern to which "first file" is returned in the event of multiple matches.

kbwestfall replied:
Very good point!  Done!

kbwestfall committed:
sorted glob

kbwestfall committed:
dev doc update

kbwestfall committed:
typo

kbwestfall commented:
Following up on the call yesterday, I've updated the dev doc and added a section discussing how to do change logging now:
```
Logging changes
~~~~~~~~~~~~~~~

It is important to log changes made to the code in a way that other developers
and eventually users can interpret.  In the past we have done this using the
single ``CHANGES.rst`` file; however, we now have version specific change logs
in the ``doc/releases`` directory.  In terms of development guidelines:

- Changes made to the code should be logged in the relevant development log.
  For example, all changes made *after* version 1.14.0 will be logged in a
  ``doc/release/1.14.1dev.rst`` file.  If the relevant file doesn't exist when
  you submit your PR, create it.

- Changes are expected to fall under a small set of broad categories, like
  improvements to performance for specific instruments, minor bug fixes, or
  datamodel changes (see previous release docs for examples).  When including
  your change, add it below the relevant heading; if no relevant heading
  exists, add a new one.

- Hotfixes merged directly to the ``release`` branch should *also be added to
  the relevant development log*.  I.e., these changes are not part of the
  released tag, even if they are in the "release" branch.  Again, if the
  relevant file doesn't exist when you perform the hotfix, create it in a way
  that it will get merged with the identical doc in the ``develop`` branch.

- When tagging, the development log will be renamed to the new tag version, and
  a new file should be created for the next development phase.  See
  :ref:`tagging`.
```

Let me know if you have any questions or additions.

kbwestfall committed:
Merge branch 'develop' into filesearch

rcooke-ast commented on CHANGES.rst:
```diff
@@ -1,3 +1,10 @@
+1.14.1dev
+---------
+
+PLEASE ADD TO doc/releases/1.14.1dev.rst!
```
Great idea!

rcooke-ast commented on doc/dev/development.rst:
```diff
@@ -130,6 +123,38 @@ Our primary means of **communication** for development is the `PypeIt developers
 Slack <https://pypeit.slack.com>`_ and a biweekly telecon.  Contact `X
 Prochaska`_ for Slack access and/or the relevant Zoom link.
 
+
```
I'm not sure where's most appropriate, but should we add something that states that it's a good idea to build and test the docs using sphinx to ensure the docstrings appear as you've intended?

kbwestfall replied:
It's mentioned in the "Workflow" section.  I added a brief comment about checking the formatting of the docs after they're built.

rcooke-ast commented on pypeit/spectrographs/keck_kcwi.py:
```diff
@@ -1282,14 +1281,13 @@ def get_rawimage(self, raw_file, det):
-        if len(fil) != 1:
-            msgs.error("Found {:d} files matching {:s}".format(len(fil), raw_file))
+        fil = utils.find_single_file(f'{raw_file}*')
+        if fil is None:
```
Should this if statement be located in `utils.find_single_file`?

kbwestfall replied:
Good idea.  There are pieces of the code that use the fact that the function returns None for files that don't exist.  So I added a keyword to indicate that the code should fault if it can't find a relevant file.

rcooke-ast reviewed (approved):
All good from me - just some minor comments to think about, and then merge when you're ready!

kbwestfall committed:
add required argument to find_single_file

kbwestfall committed:
comment on checking the doc formatting
